File: MvvmSample/app/src/main/java/com/abben/mvvmsample/ui/movies/MoviesViewModel.java
package com.abben.mvvmsample.ui.movies;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.support.annotation.NonNull;

import com.abben.mvvmsample.base.BaseAndroidViewModel;
import com.abben.mvvmsample.bean.Movie;
import com.abben.mvvmsample.db.MoviesDataBase;
import com.abben.mvvmsample.util.AbsentLiveData;
import com.abben.mvvmsample.util.Objects;
import com.abben.mvvmsample.vo.Resource;
import com.abben.mvvmsample.vo.TypeMovies;

import java.util.List;

/**
 * Created by abben on 2017/10/20.
 */
public class MoviesViewModel extends BaseAndroidViewModel {

    private final MutableLiveData<String> moveType = new MutableLiveData<>();

    private LiveData<Resource<List<Movie>>> mObservableMovies;

    private MoviesRepository moviesRepository;

    public MoviesViewModel(@NonNull Application application) {
        super(application);
        MoviesDataBase moviesDataBase = MoviesDataBase.getsInstance(application);
        moviesRepository = new MoviesRepository(moviesDataBase);

        mObservableMovies = Transformations.switchMap(moveType, input -> {
            if (input == null) {
                return AbsentLiveData.create();
            } else {
                return moviesRepository.getMovies(input);
            }
        });
    }

    public void setMoveType(@TypeMovies.TypeMoviesAnnotation String moveType) {
        if (Objects.equals(this.moveType.getValue(), moveType)) {
            return;
        }
        this.moveType.setValue(moveType);
    }

    public void retry() {
        if (this.moveType.getValue() != null) {
            this.moveType.setValue(this.moveType.getValue());
        }
    }

    public LiveData<Resource<List<Movie>>> getmObservableMovies() {
        return mObservableMovies;
    }


    public LiveData<Boolean> getmRefreshing() {
        return moviesRepository.getRefreshing();
    }
}

File: MvvmSampleJetpack/app/src/main/java/com/abben/mvvmsamplejetpack/ui/movies/MoviesViewModel.kt
package com.abben.mvvmsamplejetpack.ui.movies

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.abben.mvvmsamplejetpack.base.BaseAndroidViewModel
import com.abben.mvvmsamplejetpack.bean.Movie
import com.abben.mvvmsamplejetpack.db.MoviesDataBase
import com.abben.mvvmsamplejetpack.vo.Resource

/**
 *  @author abben
 *  on 2019/1/21
 */
class MoviesViewModel(baseApplication: Application) : BaseAndroidViewModel(baseApplication) {

    val moveType: MutableLiveData<String> = MutableLiveData()

    var mObservableMovies: LiveData<Resource<List<Movie>>>? = null

    private var moviesRepository: MoviesRepository

    init {
        val moviesDataBase = MoviesDataBase.getsInstance(baseApplication)
        moviesRepository = MoviesRepository(moviesDataBase)
        mObservableMovies = Transformations.switchMap(moveType) {
            moviesRepository.getMovies(it)
        }
    }

    fun retry() {
        moveType.value?.let {
            moveType.value = moveType.value
        }
    }

    fun getmRefreshing(): LiveData<Boolean> {
        return moviesRepository.mMoviesRefreshing
    }
}